language/src/typechecker.rs

```diff
@@ -504,7 +501,11 @@ fn bind_variable_type(
     }
 }
 
-type VarContext = HashMap<usize, (Typ, String)>;
+#[derive(Clone)]
+struct VarContext {
+    vars: HashMap<usize, (Typ, String, bool)>,
+    mutable_vars: HashMap<usize, (Typ, String, bool)>,
+}
```

Comment:
Maybe you could add some method on this, so that we have less duplicated code, i.e.:

```rust
impl VarContext {
    fn new() -> Self {
        VarContext {
            vars: HashMap::new(),
            mutable_vars: HashMap::new(),
        }
    }
}
```

Reply:
Yes, I have added some functions, to make the code more readable (and easier to update), and less error prone.